src/crypto/tls/ech.go

```diff
@@ -0,0 +1,818 @@
+			// yield significant savings. However, it's useful to test that our
+			// server correctly handles a sequence of compressed extensions and
+			// not just one.
+			outerExtensions = []uint16{
+				extensionStatusRequest,
+				extensionSupportedCurves,
+				extensionSupportedPoints,
+			}
+		}
+		rawCompressedHello, ok := echIncorporateOuterExtensions(helloInner.marshal(), innerDigest, outerExtensions)
+		if !ok {
+			return nil, nil, errors.New("tls: ech: compression of ClientHelloInner failed")
+		}
+
+		// Set "encrypted_client_hello".
+		var ech echClient
+		ech.handle.suite = ctx.cipherSuite()
+		ech.handle.configId = ctx.configId(echConfig.raw)
+		ech.handle.enc = enc
+		ech.payload = ctx.encrypt(rawCompressedHello)
+		hello.encryptedClientHelloOffered = true
+		hello.encryptedClientHello = ech.marshal()
+
+		// Update the HRR pre-shared-key. This is used to encrypt the second
+		// ClientHelloInner in case the server sends an HRR.
+		c.ech.hrrPsk = ctx.hrrPsk()
+
+		// Record the "random" sent in the ClientHelloInner. This value will be
+		// used for the second ClientHelloInner in case the server sends an HRR.
+		c.ech.hrrInnerRandom = helloInner.random
+
+		// Offer ECH.
+		c.ech.status.Offered = true
+		helloInner.raw = nil
+		hello.raw = nil
+		return hello, helloInner, nil
+	}
+
+	hello = new(clientHelloMsg)
+	*hello = *helloBase
+
+	hpkeSuite, err := hpkeAssembleCipherSuite(dummyKemId, dummyKdfId, dummyAeadId)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	pk, err := hpkeSuite.KEM.Deserialize(echDummyPublicKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	var dummyEch echClient
+	if !c.hrrTriggered {
+		dummyEch.handle.suite.kdfId = dummyKdfId
+		dummyEch.handle.suite.aeadId = dummyAeadId
+		dummyEch.handle.configId = make([]byte, dummyKdfOutputLen)
+		if _, err = io.ReadFull(config.rand(), dummyEch.handle.configId); err != nil {
+			return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+		}
+		c.ech.hrrCipherSuite = dummyEch.handle.suite
+		c.ech.hrrConfigId = dummyEch.handle.configId
+	} else {
+		// After HRR, the server checks these fields match.
+		dummyEch.handle.suite = c.ech.hrrCipherSuite
+		dummyEch.handle.configId = c.ech.hrrConfigId
+	}
+
+	dummyEch.handle.enc, _, err = hpke.SetupBaseS(hpkeSuite, config.rand(), pk, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	// TODO(cjpatton): Ensure the length of the dummy ciphertext doesn't stick
+	// out.
+	dummyEch.payload = make([]byte, len(hello.marshal())+dummyAeadOverheadLen)
+	if _, err = io.ReadFull(config.rand(), dummyEch.payload); err != nil {
+		return nil, nil, fmt.Errorf("tls: ech covertext: %s", err)
+	}
+
+	hello.encryptedClientHelloOffered = true
+	hello.encryptedClientHello = dummyEch.marshal()
+
+	// Bypass ECH and provide covertext.
+	c.ech.status.Grease = true
+	hello.raw = nil
+	return hello, nil, nil
+}
+
```

Comment:
@chris-wood, please re-review this function now that the server flow has changed. As a reminder, what has changed is that the server no longer detects GREASE based on the outer SNI. Instead, it presumes GREASE until the client signals that ECH was rejected ("ech_required").